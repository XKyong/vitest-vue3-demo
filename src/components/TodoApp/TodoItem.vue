<template>
  <li data-testid="todo-item" :class="{ 'completed': todo.done, 'editing': isEditing }">
    <div class="view">
      <input v-model="todo.done" data-testid="todo-done" class="toggle" type="checkbox" />
      <label data-testid="todo-text" @dblclick="isEditing = true">{{ todo.text }}</label>
      <button data-testid="delete" class="destroy" @click="$emit('delete-todo', todo.id)"></button>
    </div>
    <input v-focus="todo.text" data-testid="todo-edit" class="edit" :value="todo.text" @blur="isEditing = false" @keyup.enter="handleEditTodo" />
  </li>
</template>
<script setup lang="ts">
import { ref, type DirectiveBinding } from 'vue'
interface TodoItem {
  id: number
  text: string
  done: boolean
}

interface Props {
  todo: TodoItem
}

const props = withDefaults(defineProps<Props>(), {
  todo: () => ({} as TodoItem),
})

const emit = defineEmits<{
  (e: 'delete-todo', id: number): void
  (e: 'edit-todo', data: { id: number; text: string }): void
}>()

const isEditing = ref(false)

const vFocus = (el: HTMLInputElement, binding: DirectiveBinding<string>) => {
  if (binding.value) {
    el.focus()
  }
}

function handleEditTodo(e: KeyboardEvent) {
  emit('edit-todo', {
    id: props.todo.id,
    text: (e.target as HTMLInputElement).value
  })
}
</script>
