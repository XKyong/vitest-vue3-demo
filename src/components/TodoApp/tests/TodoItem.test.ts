import { shallowMount } from '@vue/test-utils'
import TodoItem from '../TodoItem.vue'

import type { VueWrapper } from '@vue/test-utils'

describe('TodoItem.vue', () => {
  let wrapper: VueWrapper | null = null
  const todo = { id: 1, text: '测试', done: true }

  beforeEach(() => {
    wrapper = shallowMount(TodoItem, {
      props: { todo }
    })
  })

  it('text', () => {
    expect(wrapper!.find('[data-testid="todo-text"]').text()).toBe(todo.text)
  })

  it('done', async () => {
    const done = wrapper!.find('[data-testid="todo-done"]')
    const todoItem = wrapper!.find('[data-testid="todo-item"]')
    expect((done.element as HTMLInputElement).checked).toBeTruthy()
    expect(todoItem.classes()).toContain('completed')

    await done.setValue(false)
    expect(todoItem.classes()).not.toContain('completed')

    await done.setValue(true)
    expect(todoItem.classes()).toContain('completed')
  })

  it('delete todo', async () => {
    const deleteBtn = wrapper!.find('[data-testid="delete"]')
    await deleteBtn.trigger('click')
    expect(wrapper!.emitted()['delete-todo']).toBeTruthy()
    expect((wrapper!.emitted()['delete-todo'] as any[])[0][0]).toBe(todo.id)
  })

  it('edit todo', async () => {
    const label = wrapper!.find('[data-testid="todo-text"]')
    const todoItem = wrapper!.find('[data-testid="todo-item"]')
    const todoEdit = wrapper!.find('[data-testid="todo-edit"]')
    await label.trigger('dblclick')
    expect(todoItem.classes()).toContain('editing')

    await todoEdit.trigger('blur')
    expect(todoItem.classes()).not.toContain('editing')
  })

  it('save edit todo', async () => {
    const label = wrapper!.find('[data-testid="todo-text"]')
    const todoEdit = wrapper!.find('[data-testid="todo-edit"]')
    await label.trigger('dblclick')
    
    // 编辑文本框中的内容展示
    expect((todoEdit.element as HTMLInputElement).value).toBe((wrapper!.vm as InstanceType<typeof TodoItem>).todo.text)

    const text = 'hello'
    await todoEdit.setValue(text)

    // 触发回车保存事件
    await todoEdit.trigger('keyup.enter')

    // 数据被修改了，向父组件发布事件
    // expect((wrapper!.vm as InstanceType<typeof TodoItem>).todo.text).toBe(text)
    expect(wrapper!.emitted()['edit-todo']).toBeTruthy()
    expect((wrapper!.emitted()['edit-todo'] as  any[])[0][0]).toEqual({
      id: (wrapper!.vm as InstanceType<typeof TodoItem>).todo.id,
      text
    })
  })
})
