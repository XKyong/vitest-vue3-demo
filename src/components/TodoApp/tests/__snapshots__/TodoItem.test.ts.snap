// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoItem.vue > snapshot 1`] = `
"<li data-testid="todo-item" class="completed">
  <div class="view"><input data-testid="todo-done" class="toggle" type="checkbox"><label data-testid="todo-text">测试</label><button data-testid="delete" class="destroy"></button></div><input data-testid="todo-edit" class="edit" value="测试">
</li>"
`;
